streams/src/main/java/org/apache/kafka/streams/kstream/internals/AbstractKStreamTimeWindowAggregateProcessor.java

```diff
@@ -0,0 +1,170 @@
+import org.apache.kafka.streams.state.KeyValueIterator;
+import org.apache.kafka.streams.state.TimestampedWindowStore;
+import org.apache.kafka.streams.state.ValueAndTimestamp;
+
+public abstract class AbstractKStreamTimeWindowAggregateProcessor<KIn, VIn, VAgg> extends ContextualProcessor<KIn, VIn, Windowed<KIn>, Change<VAgg>> {
+
+    private final Time time = Time.SYSTEM;
+    private final String storeName;
+    private final EmitStrategy emitStrategy;
+    private final boolean sendOldValues;
+    protected final TimeTracker timeTracker = new TimeTracker();
+
+    private TimestampedTupleForwarder<Windowed<KIn>, VAgg> tupleForwarder;
+    protected TimestampedWindowStore<KIn, VAgg> windowStore;
+    protected Sensor droppedRecordsSensor;
+    protected Sensor emittedRecordsSensor;
+    protected Sensor emitFinalLatencySensor;
+    protected long lastEmitWindowCloseTime = ConsumerRecord.NO_TIMESTAMP;
+    protected InternalProcessorContext<Windowed<KIn>, Change<VAgg>> internalProcessorContext;
+
+    protected AbstractKStreamTimeWindowAggregateProcessor(final String storeName,
+                                                          final EmitStrategy emitStrategy,
+                                                          final boolean sendOldValues) {
+        this.storeName = storeName;
+        this.emitStrategy = emitStrategy;
+        this.sendOldValues = sendOldValues;
+    }
+
+    @Override
+    public void init(final ProcessorContext<Windowed<KIn>, Change<VAgg>> context) {
+        super.init(context);
+        internalProcessorContext = (InternalProcessorContext<Windowed<KIn>, Change<VAgg>>) context;
+        final StreamsMetricsImpl metrics = internalProcessorContext.metrics();
+        final String threadId = Thread.currentThread().getName();
+        droppedRecordsSensor = droppedRecordsSensor(threadId, context.taskId().toString(), metrics);
+        emittedRecordsSensor = emittedRecordsSensor(threadId, context.taskId().toString(),
+            internalProcessorContext.currentNode().name(), metrics);
+        emitFinalLatencySensor = emitFinalLatencySensor(threadId, context.taskId().toString(),
+            internalProcessorContext.currentNode().name(), metrics);
+        windowStore = context.getStateStore(storeName);
+
+        if (emitStrategy.type() == StrategyType.ON_WINDOW_CLOSE) {
+            // Restore last emit close time for ON_WINDOW_CLOSE strategy
+            final Long lastEmitTime = internalProcessorContext.processorMetadataForKey(storeName);
+            if (lastEmitTime != null) {
+                lastEmitWindowCloseTime = lastEmitTime;
+            }
+            final long emitInterval = StreamsConfig.InternalConfig.getLong(
+                context.appConfigs(),
+                EMIT_INTERVAL_MS_KSTREAMS_WINDOWED_AGGREGATION,
+                1000L
+            );
+            timeTracker.setEmitInterval(emitInterval);
+        } else {
+            tupleForwarder = new TimestampedTupleForwarder<>(
```

Comment:
In the `emit on final` case, the tuple forwarder would not be used (you can see that in the old code, we have a separate constructor for it in `emit on final` which does not pass in the cache at all) since we always rely on the logic to scan/emit to downstream that does not call the tuple forwarder. I think it's cleaner to not construct the object at all for this case.

Reply:
EDIT: actually you're right! :) The maybeForward should still be called in `emit final` case.